File: sdk/ts/src/resolution/forwardResolver.ts
import { PublicKey } from '@metaplex-foundation/umi';
import { parseWalletTuple, normalizeChainCaip2 } from './caip';
import { DEFAULT_SOLANA_CAIP2, SRS_DEFAULT_PROGRAM_ID } from './constants';
import { namehash } from './namehash';
import { findRecordPda } from './pda';
import type { RecordAccountProvider } from './provider';
import {
  parseResolutionTuples,
  type ResolutionTuple,
} from './tupleCodec';

export interface DomaForwardResolverConfig {
  provider: RecordAccountProvider;
  domaClassAddress: PublicKey;
  programId?: PublicKey;
  defaultChainCaip2?: string;
}

export class DomaForwardResolver {
  private readonly provider: RecordAccountProvider;
  private readonly domaClassAddress: PublicKey;
  private readonly programId: PublicKey;
  private readonly defaultChainCaip2: string;

  constructor(config: DomaForwardResolverConfig) {
    this.provider = config.provider;
    this.domaClassAddress = config.domaClassAddress;
    this.programId = config.programId ?? SRS_DEFAULT_PROGRAM_ID;
    this.defaultChainCaip2 = normalizeChainCaip2(
      config.defaultChainCaip2 ?? DEFAULT_SOLANA_CAIP2
    );
  }

  async resolve(
    name: string,
    chainCaip2: string = this.defaultChainCaip2
  ): Promise<string | null> {
    const normalizedChain = normalizeChainCaip2(chainCaip2);
    const tuples = await this.fetchNameTuples(name);
    if (!tuples) {
      return null;
    }

    let selectedWallet: string | null = null;
    for (const [key, value] of tuples) {
      const parsed = parseWalletTuple(key, value, normalizedChain);
      if (!parsed) {
        continue;
      }

      if (parsed.chainId === normalizedChain) {
        selectedWallet = parsed.walletAddress;
      }
    }

    return selectedWallet;
  }

  async resolveRecord(name: string, recordKey: string): Promise<string | null> {
    const tuples = await this.fetchNameTuples(name);
    if (!tuples) {
      return null;
    }

    const normalizedKey = recordKey.trim().toUpperCase();
    if (normalizedKey.length === 0) {
      return null;
    }

    let selectedValue: string | null = null;
    for (const [key, value] of tuples) {
      if (key.trim().toUpperCase() !== normalizedKey) {
        continue;
      }

      const candidate = value.trim();
      if (candidate.length > 0) {
        selectedValue = candidate;
      }
    }

    return selectedValue;
  }

  async resolveRecords(name: string, recordKey: string): Promise<string[]> {
    const tuples = await this.fetchNameTuples(name);
    if (!tuples) {
      return [];
    }

    const normalizedKey = recordKey.trim().toUpperCase();
    if (normalizedKey.length === 0) {
      return [];
    }

    return tuples
      .filter(([key]) => key.trim().toUpperCase() === normalizedKey)
      .map(([, value]) => value.trim())
      .filter((value) => value.length > 0);
  }

  private async fetchNameTuples(name: string): Promise<ResolutionTuple[] | null> {
    const tokenId = namehash(name);
    const [recordPda] = await findRecordPda(
      this.domaClassAddress,
      tokenId,
      this.programId
    );

    return this.fetchRecordTuples(recordPda);
  }

  private async fetchRecordTuples(recordPda: PublicKey) {
    const record = await this.provider.fetchRecord(recordPda);
    if (!record) {
      return null;
    }

    return parseResolutionTuples(record.data);
  }
}

File: sdk/rust/src/resolution/forward_resolver.rs
use std::sync::Arc;

use solana_program::pubkey::Pubkey;

use crate::resolution::caip::{normalize_chain_caip2, parse_wallet_tuple};
use crate::resolution::constants::{DEFAULT_SOLANA_CAIP2, SRS_DEFAULT_PROGRAM_ID};
use crate::resolution::errors::ResolutionResult;
use crate::resolution::namehash::namehash;
use crate::resolution::pda::find_record_pda;
use crate::resolution::provider::RawRecordAccountProvider;
use crate::resolution::record_decoder::decode_srs_record;
use crate::resolution::tuple_codec::{parse_resolution_tuples, ResolutionTuple};

pub trait ForwardNameResolver: Send + Sync {
    fn resolve_forward(
        &self,
        name: &str,
        chain_caip2: Option<&str>,
    ) -> ResolutionResult<Option<String>>;
}

pub struct DomaForwardResolverConfig {
    pub provider: Arc<dyn RawRecordAccountProvider>,
    pub doma_class_address: Pubkey,
    pub program_id: Option<Pubkey>,
    pub default_chain_caip2: Option<String>,
}

#[derive(Clone)]
pub struct DomaForwardResolver {
    provider: Arc<dyn RawRecordAccountProvider>,
    doma_class_address: Pubkey,
    program_id: Pubkey,
    default_chain_caip2: String,
}

impl DomaForwardResolver {
    pub fn new(config: DomaForwardResolverConfig) -> ResolutionResult<Self> {
        let default_chain_caip2 = normalize_chain_caip2(
            config
                .default_chain_caip2
                .as_deref()
                .unwrap_or(DEFAULT_SOLANA_CAIP2),
        )?;

        Ok(Self {
            provider: config.provider,
            doma_class_address: config.doma_class_address,
            program_id: config.program_id.unwrap_or(SRS_DEFAULT_PROGRAM_ID),
            default_chain_caip2,
        })
    }

    pub fn resolve(
        &self,
        name: &str,
        chain_caip2: Option<&str>,
    ) -> ResolutionResult<Option<String>> {
        let normalized_chain =
            normalize_chain_caip2(chain_caip2.unwrap_or(&self.default_chain_caip2))?;
        let token_id = namehash(name)?;
        let (record_pda, _) =
            find_record_pda(&self.doma_class_address, &token_id, &self.program_id)?;

        let Some(tuples) = self.fetch_record_tuples(&record_pda)? else {
            return Ok(None);
        };

        let mut selected_wallet: Option<String> = None;
        for (key, value) in tuples {
            let Some(parsed) = parse_wallet_tuple(&key, &value, &normalized_chain)? else {
                continue;
            };

            if parsed.chain_id == normalized_chain {
                selected_wallet = Some(parsed.wallet_address);
            }
        }

        Ok(selected_wallet)
    }

    fn fetch_record_tuples(
        &self,
        record_pda: &Pubkey,
    ) -> ResolutionResult<Option<Vec<ResolutionTuple>>> {
        let Some(raw) = self.provider.fetch_raw_record_account(record_pda)? else {
            return Ok(None);
        };

        let decoded = decode_srs_record(&raw)?;
        Ok(Some(parse_resolution_tuples(&decoded.data)?))
    }
}

impl ForwardNameResolver for DomaForwardResolver {
    fn resolve_forward(
        &self,
        name: &str,
        chain_caip2: Option<&str>,
    ) -> ResolutionResult<Option<String>> {
        self.resolve(name, chain_caip2)
    }
}
